### PR TITLE
Support junit 5; Migrate non-forked junit 3 test cases

### PR DESCRIPTION
**Pull request by Vlatombe**
Migrate existing junit 3 tests to junit 5 with standard `ParameterizedTest` facility

This uses `@MethodSource` providing the various ChannelRunner instances used for testing.

Migrated existing assert/assume calls impacted tests to the new jupiter api.

Overall this covers more combinations than the previous framework as some of them were actually not covered previously for some reason, even though they were extending `RmiTestBase`.

Keeping existing junit 3 test cases from forked packages as the proper approach here would rather be to pull new versions.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by jglick** [comment]
https://github.com/jenkinsci/remoting/pull/593/checks?check_run_id=8873100454 flake or real?

File: src/main/java/hudson/remoting/Checksum.java [hunk 1]
```diff
@@ -19,7 +20,7 @@
  *
  * @author Kohsuke Kawaguchi
  */
-final class Checksum {
+final class Checksum implements Serializable {
```
Comment (jglick) [inline src/main/java/hudson/remoting/Checksum.java:23]: ?

File: src/test/java/hudson/remoting/ChannelRunners.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.stream.Stream;
+
+public final class ChannelRunners {
+    public static final String PROVIDER_METHOD = "hudson.remoting.ChannelRunners#provider";
+
+    private ChannelRunners() {}
+
+    @SuppressWarnings("unused") // used by JUnit
+    static Stream<ChannelRunner> provider() {
```
Comment (jglick) [inline src/test/java/hudson/remoting/ChannelRunners.java:11]:
```suggestion
    public static Stream<ChannelRunner> provider() {
```
? Seems weird to be accessing a non`public` method reflectively.

File: src/main/java/hudson/remoting/Checksum.java (hunk 1 above)
Comment (Vlatombe) [reply src/main/java/hudson/remoting/Checksum.java:23]: Comes from one of the combination of `hudson.remoting.PrefetchingTest`, which I guess was not tested before.

**Comment by Vlatombe** [comment]
> [#593 (checks)](https://github.com/jenkinsci/remoting/pull/593/checks?check_run_id=8873100454) flake or real?

~50% flake locally if I recall properly.

File: src/test/java/hudson/remoting/DeadRemoteOutputStreamTest.java [hunk 3]
```diff
@@ -55,7 +64,7 @@ public Void call() throws Exception {
             StringWriter sw = new StringWriter();
             e.printStackTrace(new PrintWriter(sw));
             String whole = sw.toString();
-            assertTrue(whole, whole.contains(MESSAGE) && whole.contains("hudson.rem0ting.TestCallable"));
+            assertTrue(whole.contains(MESSAGE) && whole.contains("hudson.rem0ting.TestCallable"), whole);
```
Comment (jglick) [inline src/test/java/hudson/remoting/DeadRemoteOutputStreamTest.java:67]: Or switch to Hamcrest assertions.

File: src/test/java/hudson/remoting/PipeTest.java [hunk 4]
```diff
@@ -68,15 +77,19 @@ public void testRemoteWrite() throws Exception {
...
+        channelRunner.withChannel(channel -> {
+            final Pipe p = Pipe.createRemoteToLocal();
+            final Future<Void> f = channel.callAsync(new InfiniteWriter(p));
+            try (InputStream in = p.getIn()) {
+                assertEquals(in.read(), 0);
+            }
 
-        final ExecutionException e = assertThrows(ExecutionException.class, f::get);
-        assertThat(e.getCause(), instanceOf(IOException.class));
+            final ExecutionException e = assertThrows(ExecutionException.class, () -> f.get(2, TimeUnit.SECONDS));
```
Comment (jglick) [inline src/test/java/hudson/remoting/PipeTest.java:90]: Correcting flake I guess?

File: src/test/java/hudson/remoting/PipeWriterTest.java [hunk 5]
```diff
@@ -19,12 +25,11 @@ public class PipeWriterTest extends RmiTestBase implements Serializable, PipeWri
      * Proxy that can be used from the other side to verify the state.
      */
     PipeWriterTestChecker checker;
-
-    @Override
-    protected void setUp() throws Exception {
-        super.setUp();
-        // Checker operates using the user-space RMI
```
Comment (jglick) [inline src/test/java/hudson/remoting/PipeWriterTest.java:26]: lost comment

**Review by jglick — APPROVED** [review]
OK I think. JUnit 5 is pretty opaque to me.

Does this patch make Surefire timeouts apply automatically?

File: src/test/java/hudson/remoting/PipeTest.java (hunk 4 above)
Comment (Vlatombe) [reply src/test/java/hudson/remoting/PipeTest.java:90]: Yes

**Comment by jglick** [comment]
Seems `ChannelTest` can still hang?

**Comment by jglick** [comment]
Do you plan to look into https://github.com/jenkinsci/remoting/pull/593#issuecomment-1278857617, or shall I just merge this as is?

**Comment by Vlatombe** [comment]
Added a timeout, I don't plan to investigate more time into the hanging test.

**Comment by jglick** [comment]
https://github.com/jenkinsci/remoting/pull/593/commits/ca83285cc87ec1870a9bcb25232fd2a06cac2a50 did not work it seems:

```
00:06:57.224  [INFO] Running hudson.remoting.ChannelTest
00:17:11.572  Sending interrupt signal to process
```

**Comment by jglick** [comment]
Maybe it worked so far as it went, but each test method gets run up to five times, and there are up to six parameter values, so spending a minute on each run in the worst case can be a lot.

Hmm, but it seems that by default timeouts are not preëmptive—if you exceed the timeout (but the method completes), the test fails, but that is no good for ensuring CI completes in time.

**Comment by jglick** [comment]
Some really weird stuff like https://github.com/jenkinsci/remoting/blob/ca83285cc87ec1870a9bcb25232fd2a06cac2a50/src/test/java/org/jenkinsci/remoting/protocol/impl/ConnectionHeadersFilterLayerTest.java#L68-L71

**Pull request merged**